### PR TITLE
Weights prefs

### PR DESCRIPTION
**Pull request by piatra**
As described in #1101 

@sarracini || @k88hudson please review.

---
This change is [Reviewable](https://reviewable.io/reviews/mozilla/activity-stream/1106)

### CODE REVIEW

File: common/recommender/Baseline.js [hunk 1]
```diff
@@ -10,8 +11,9 @@ const {INFINITE_SCROLL_THRESHOLD} = require("../constants");
  * @param {Array.<URLs>} history - User history used to assign higher score to popular domains.
  */
 class Baseline {
-  constructor(history) {
+  constructor(history, options = {}) {
```
Comment (sarracini) [inline common/recommender/Baseline.js]: are you able to update the @param comment with the addition of options?

File: common/recommender/Baseline.js [hunk 2]
```diff
@@ -30,11 +32,15 @@ class Baseline {
...
       // Features weights: Positive values decrease the score proportional to a factor of feature * weight.
       //                   Negative values increase score proportional to a factor of feature * weight.
-      [0.4, 0.7, 0.1, -0.4, -0.2, -0.1]);
+      this.options.highlightsCoefficients || COEFFICIENTS);
 
     return Object.assign({}, entry, {score}, {host});
   }
 
+  updateOptions(options = {}) {
+    this.options = options;
```
Comment (sarracini) [inline common/recommender/Baseline.js]: i'm assuming the only time you're ever going to call ```updateOptions``` is when you need to pass in new options - so i think it's safe to assume you'll have a param here, so you can remove ```= {}```

File: content-test/recommender/Baseline.test.js [hunk 3]
```diff
@@ -126,4 +126,20 @@ describe("Baseline", () => {
...
+    it("should instantiate the recommender with empty object by default", () => {
+      baseline = new Baseline(fakeHistory);
+      assert.deepEqual(baseline.options, {});
+    });
+    it("should instantiate the recommender with the correct settings", () => {
+      baseline = new Baseline(fakeHistory, options);
+      assert.deepEqual(baseline.options, options);
+    });
+    it("should update the correct settings", () => {
+      baseline.updateOptions(options);
```
Comment (sarracini) [inline content-test/recommender/Baseline.test.js]: you didn't actually change your options ;) 
maybe 
```js
const newOptions = {highlightsCoefficients: [4, 5, 6]};
baseline.updateOptions(newOptions);
assert.deepEqual(baseline.options, newOptions);
```

File: lib/ActivityStreams.js [hunk 4]
```diff
@@ -170,6 +176,23 @@ ActivityStreams.prototype = {
...
+   * @private
+   */
+  _loadRecommender() {
+    // Only need to load history items once per session.
+    if (this._baselineRecommender !== null) {
+      return;
+    }
+
+    this._memoized.getAllHistoryItems().then(historyItems => {
+      let highlightsCoefficients = JSON.parse(simplePrefs.prefs.weightedHighlightsCoefficients);
```
Comment (sarracini) [inline lib/ActivityStreams.js]: 👍 🎉

File: lib/ActivityStreams.js [hunk 5]
```diff
@@ -411,6 +432,17 @@ ActivityStreams.prototype = {
       this._prefsProvider.actionHandler(args);
     };
     this.on(CONTENT_TO_ADDON, this._contentToAddonHandlers);
+
+    simplePrefs.on("", prefName => {
```
Comment (sarracini) [inline lib/ActivityStreams.js]: hmm is it possible use the PrefsProvider to do this? that way we have a communal place where the prefs are being changed? @k88hudson thoughts?

Comment (k88hudson) [inline lib/ActivityStreams.js]: Right now there isn't really a convenient way to use `PrefsProvider` for this use case – it's more for handling communication with content, although we could add something to do it (i.e. an `addListener`/`on` method) – the advantage of that would be we could handle removing listeners on unload in once place.

For now, this seems reasonable (we could file the enhancement as separate ticket) -- although this method needs to be cached on the instance and removed at unload time.

File: common/recommender/Baseline.js [hunk 6]
```diff
@@ -2,6 +2,7 @@
 
 const URL = require("../../lib/vendor.bundle.js").urlParse;
 const {INFINITE_SCROLL_THRESHOLD} = require("../constants");
+const COEFFICIENTS = [0.4, 0.7, 0.1, -0.4, -0.2, -0.1];
```
Comment (pdehaan) [inline common/recommender/Baseline.js]: This seems to be the same values as in [package.json](https://github.com/mozilla/activity-stream/pull/1106/files#diff-b9cfc7f2cdf78a7f4b91a753d10865a2R185) edits below:

```js
 +    {
 +      "name": "weightedHighlightsCoefficients",
 +      "title": "recommender weights",
 +      "description": "weightedHighlightsCoefficients",
 +      "type": "string",
 +      "value": "[0.4, 0.7, 0.1, -0.4, -0.2, -0.1]",
 +      "hidden": true
      }
```

Not sure if there is a way to consolidate the values so they don't get out of sync if somebody tweaks these values in the future.

File: content-test/recommender/Baseline.test.js (hunk 3 above)
Comment (piatra) [inline content-test/recommender/Baseline.test.js]: The beforeEach created a new baseline for every `it` block which by default has no options. If the assert passes it means it got the new options.

File: common/recommender/Baseline.js (hunk 6 above)
Comment (piatra) [inline common/recommender/Baseline.js]: @pdehaan you can load the package.json and read the values. Not sure if that's a good practice™®

File: lib/ActivityStreams.js (hunk 5 above)
Comment (piatra) [inline lib/ActivityStreams.js]: /note to self: unload method.

File: lib/ActivityStreams.js (hunk 4 above)
Comment (sarracini) [inline lib/ActivityStreams.js]: let's wrap this in a try/catch to handle faulty pref values and then make the constructor fallback to defaults

**Comment by sarracini** [comment]
🎉  R+ from me!

**Pull request merged**